// build-watcher/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require('path');
var webpack = require('webpack');
var getWebpackDevConfig = require('./webpack.dev.config');
var rx_1 = require("rx");
var BuildWatcher = (function () {
    function BuildWatcher(presentationPath) {
        this.presentationPath = presentationPath;
    }
    BuildWatcher.prototype.start = function () {
        if (!this.watcher$) {
            var watcher = this.createWatcher({
                ignored: ["**/*.*", "!**/*.vue", "!**/structure.json"]
            });
            return rx_1.Observable.just(this.presentationPath)
                .flatMap(getWebpackDevConfig)
                .map(webpack)
                .map(watcher)
                .mergeAll();
        }
        return this.watcher$;
    };
    BuildWatcher.prototype.stop = function () {
        return rx_1.Observable.fromCallback(this.watcher.close, this.watcher)()
            .toPromise();
    };
    BuildWatcher.prototype.createWatcher = function (options) {
        var _this = this;
        return function (compiler) {
            var watch = function () {
                return this.watcher = compiler.watch.apply(compiler, arguments);
            };
            return _this.watcher$ = rx_1.Observable.fromNodeCallback(watch, _this)(options);
        };
    };
    return BuildWatcher;
}());
exports.BuildWatcher = BuildWatcher;
