// build-watcher/webpack.common.config.js
const path = require('path');
const ProgressBarPlugin = require('progress-bar-webpack-plugin');
const VueLoaderOptionsPlugin = require('vue-loader-options-plugin');

const babelPresets = [
  require.resolve('babel-preset-es2015'),
  require.resolve('babel-preset-stage-0'),
];

module.exports = function (structure, presentationPath) {
  return {
    resolveLoader: {
      modules: [
        path.join(require.resolve(__dirname), '../..'),
        path.join(__dirname, 'node_modules'),
        path.join(__dirname, 'loaders'),
      ],
    },
    resolve: {
      extensions: ['.webpack.js', '.web.js', '.ts', '.tsx', '.js'],
      alias: {
        vue: 'vue/dist/vue.esm.js',
      },
    },
    module: {
      loaders: [
        {
          test: /\.ts(x?)$/,
          loader: 'awesome-typescript-loader',
          options: {
            silent: true,
          },
        },
        {
          test: /\.vue$/,
          loader: 'slide-template-loader',
          query: {
            structure,
            presentationPath
          },
        },
        {
          test: /\.vue$/,
          loader: 'vue-loader',
          options: {
            loaders: {
              i18n: '@kazupon/vue-i18n-loader',
              css: 'style-loader!css-loader!autoprefixer-loader',
            },
          },
        },
        {
          test: /\.css$/,
          exclude: /node_modules/,
          loader: 'style-loader!css-loader!autoprefixer-loader',
        },
        {
          test: /\.png$/,
          exclude: /node_modules/,
          loader: 'url-loader',
        },
        {
          test: /\.(ttf|eot|otf|svg|woff(2)?)(\?[a-z0-9=&.]+)?$/,
          exclude: /node_modules/,
          loader: 'url-loader',
        },
        {
          test: /\.js$/,
          exclude: /node_modules/,
          loader: 'babel-loader',
          query: {
            presets: babelPresets,
          },
        },
        {
          test: /node_modules\/routie\/dist\/routie\.js$/,
          exclude: /node_modules/,
          loader: 'imports-loader?define=>false',
        },
      ],
    },
    plugins: [
      new ProgressBarPlugin(),
      new VueLoaderOptionsPlugin({
        babel: {
          presets: babelPresets,
        },
      }),
    ],
  };
};
